### PR TITLE
[1.1.x] Fix TMC2208 SW serial communication by using new constructor

### PR DESCRIPTION
**Pull request by teemuatlut**
Move the software serial instance inside TMC2208Stepper instance to fix SW serial communication with multiple TMC2208 steppers.
The core issue is with the SoftwareSerial implementation and appeared such that only the last read call would return valid data. For example, having X, Y and Z TMC steppers, only the Z would return something else other than 0x0. The issue has not affected write capabilities or when HW serial was used.
The fix boils down to being able to use `serial.listen()`, but this requires the stepper instance to have direct access to the SW serial class. Can't use the parent `Stream` class like before.

---
I also took the opportunity to update the enabled default driver types.
Now all driver slots are disabled by default and Marlin will default to using the same timing parameters as with A4988.

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -530,27 +530,30 @@
...
 /**
- * Specify Stepper Driver types
- * The options are used to determine driver pulse timings as well as more advanced functionality.
- * Stepper timing options can be overridden in Configuration_adv.h
+ * Stepper Drivers
+ *
+ * These settings allow Marlin to tune stepper driver timing and enable advanced options for
+ * stepper drivers that support them. You may also override timing options in Configuration_adv.h.
+ *
+ * A4988 is assumed for unspecified drivers.
```
Comment (teemuatlut) [inline Marlin/Configuration.h:538]: @thinkyhead That's not quite right. The assumption is made only if no drivers are specified.

Comment (thinkyhead) [reply Marlin/Configuration.h:538]: Semantics. The individual driver is assumed to be A4988. The fact that the code compiles based on the slowest driver is not a detail exposed to users.

**Comment by Artefact2** [comment]
These commits introduce a regression for me. I have stealthChop disabled, but after merging this, the steppers always run with it enabled. I suspect other settings aren't being properly applied either, as the X axis now moves at half the speed (Likely that the wrong microstepping value gets applied too.)

Just in case, here are the outputs of M122 after homing before and after merging. Not sure how useful this is, considering the issue being addressed in this merge request.

Before merging. As expected the data returned for X and Y is nonsensical but E0 is correct.
~~~
>>> M122
SENDING:M122
  X Y E0
Enabled  false false false
Set current 800 800 500
RMS current 1436 1436 489
MAX current 2025 2025 689
Run current 25/31 25/31 15/31
Hold current 12/31 12/31 7/31
CS actual  0/31 0/31 7/31
PWM scale  0 0 9
vsense  0=.325 0=.325 1=.18
stealthChop false false false
msteps  256 256 32
tstep  0 0 0
pwm
threshold  0 0 0
[mm/s]  - - -
OT prewarn false false false
OT prewarn has
been triggered false false false
off time  0 0 5
blank time 16 16 24
hysteresis
-end  -3 -3 2
-start  1 1 3
Stallguard thrs
DRVSTATUS X Y E0
stst    X
olb    X
ola    X
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers: X = 0x00:00:00:00
 Y = 0x00:00:00:00
 E0 = 0x00:00:00:00
~~~

After merging. Currents and stealthChop look fine here, but don't seem to be actually applied. Microstepping values are wrong for X/Y (I use 16) but correct for E0.
~~~
>>> G28
SENDING:G28
>>> M122
SENDING:M122
  X Y E0
Enabled  false false false
Set current 800 800 500
RMS current 795 795 489
MAX current 1121 1121 689
Run current 25/31 25/31 15/31
Hold current 12/31 12/31 7/31
CS actual  7/31 7/31 7/31
PWM scale  9 9 0
vsense  1=.18 1=.18 1=.18
stealthChop false false false
msteps  32 32 32
tstep  0 0 1048575
pwm
threshold  0 0 0
[mm/s]  - - -
OT prewarn false false false
OT prewarn has
been triggered false false false
off time  5 5 5
blank time 24 24 24
hysteresis
-end  2 2 2
-start  3 3 3
Stallguard thrs
DRVSTATUS X Y E0
stst  X X X
olb
ola
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers: X = 0x80:07:00:00
 Y = 0x80:07:00:00
 E0 = 0x80:07:00:00
~~~

This could still be an issue on my end that's only revealed by these changes. I'm still investigating the problem.

**Comment by kakou-fr** [comment]
same problem, using TMC2208 on X Y Z (TMC2130 on E0), step are bad, driver register 0x00... and so my printer doesn't move correctly

```
Send: M122
Recv: 		X	Y	Z	E0
Recv: Enabled		false	false	false	false
Recv: Set current	500	500	700	550
Recv: RMS current	883	883	1215	520
Recv: MAX current	1245	1245	1713	733
Recv: Run current	15/31	15/31	21/31	16/31
Recv: Hold current	7/31	7/31	10/31	8/31
Recv: CS actual		0/31	0/31	0/31	8/31
Recv: PWM scale		0	0	0	1
Recv: vsense		0=.325	0=.325	0=.325	1=.18
Recv: stealthChop	false	false	false	true
Recv: msteps		256	256	256	16
Recv: tstep		0	0	0	1048575
Recv: pwm
Recv: threshold		79	79	329	1821
Recv: [mm/s]		1601.27	1601.27	96.12	1.00
Recv: OT prewarn	false	false	false	false
Recv: OT prewarn has
Recv: been triggered	false	false	false	false
Recv: off time		0	0	0	5
Recv: blank time	16	16	16	24
Recv: hysteresis
Recv: -end		-3	-3	-3	2
Recv: -start		1	1	1	3
Recv: Stallguard thrs				0
Recv: DRVSTATUS	X	Y	Z	E0
Recv: stallguard
Recv: sg_result					0
Recv: fsactive
Recv: stst					X
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:	X = 0x00:00:00:00
Recv: 	Y = 0x00:00:00:00
Recv: 	Z = 0x00:00:00:00
Recv: 	E0 = 0x80:08:00:00
```

switched to the previous version and it's ok (but only info for Z)
```
Send: M122
Recv: 		X	Y	Z	E0
Recv: Enabled		false	false	true	false
Recv: Set current	500	500	700	550
Recv: RMS current	883	883	673	520
Recv: MAX current	1245	1245	949	733
Recv: Run current	15/31	15/31	21/31	16/31
Recv: Hold current	7/31	7/31	10/31	8/31
Recv: CS actual		0/31	0/31	10/31	8/31
Recv: PWM scale		0	0	12	3
Recv: vsense		0=.325	0=.325	1=.18	1=.18
Recv: stealthChop	false	false	true	true
Recv: msteps		256	256	16	16
Recv: tstep		0	0	1048575	1048575
Recv: pwm
Recv: threshold		79	79	329	1821
Recv: [mm/s]		1601.27	1601.27	6.01	1.00
Recv: OT prewarn	false	false	false	false
Recv: OT prewarn has
Recv: been triggered	false	false	false	false
Recv: off time		0	0	5	5
Recv: blank time	16	16	24	24
Recv: hysteresis
Recv: -end		-3	-3	2	2
Recv: -start		1	1	3	3
Recv: Stallguard thrs				0
Recv: DRVSTATUS	X	Y	Z	E0
Recv: stallguard
Recv: sg_result					0
Recv: fsactive
Recv: stst				X	X
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:	X = 0x00:00:00:00
Recv: 	Y = 0x00:00:00:00
Recv: 	Z = 0xC0:0A:00:00
Recv: 	E0 = 0x80:08:00:00
```

**Comment by thinkyhead** [comment]
> These commits introduce a regression for me.

I can't see any changes in this PR that would have such an effect. Only one `cpp` file was altered, and it only allows use of software serial where previously only hardware serial was available.

**Comment by teemuatlut** [comment]
The change did affect which object is used for communication and there are some small differences. I don't think we'll need to fix anything Marlin side and the solution will likely be in the `TMC2208Stepper` library.
I'm looking into this right now and earlier today I got suggested a solution that _might_ fix the problem.

**Comment by teemuatlut** [comment]
Please try the latest version of the library from github.
I added a fix but would like to verify before creating a new release.

**Comment by Artefact2** [comment]
Thanks @teemuatlut for working on the issue. Using the latest library, settings seem to get applied OK, but reading the driver registers with `M122` is still a bit weird, in my instance the data for Y is wrong (msteps should be 16; the axis moves fine and is definitely using 16 microsteps).

[serial.txt](https://github.com/MarlinFirmware/Marlin/files/2307455/serial.txt)

**Comment by teemuatlut** [comment]
Could it be that you might have an issue with the communication in general?
It's quite hard for us to compare with previous versions as the SW serial communication didn't work before on all drivers.
I'm asking because every value on your Y driver is register value `0`. The `E0` hex value doesn't look good either but at least the rest of the values were retrieved correctly.

**Comment by kakou-fr** [comment]
for me, it's ok for Y and Z, but X is bad

just checked the wire with a multimeter, 1kohm on 63. connected ont 63 and 40.

exchange X and Y wire, same result (i show Y but not X)


```

Send: M122
Recv: 		X	Y	Z	E0
Recv: Enabled		false	true	true	false
Recv: Set current	500	500	600	550
Recv: RMS current	883	489	581	520
Recv: MAX current	1245	689	819	733
Recv: Run current	15/31	15/31	18/31	16/31
Recv: Hold current	7/31	7/31	9/31	8/31
Recv: CS actual		0/31	7/31	9/31	8/31
Recv: PWM scale		0	54	124	29
Recv: vsense		0=.325	1=.18	1=.18	1=.18
Recv: stealthChop	false	true	true	true
Recv: msteps		256	16	16	16
Recv: tstep		0	1048575	1048575	1048575
Recv: pwm
Recv: threshold		79	79	329	1821
Recv: [mm/s]		1601.27	100.08	6.01	1.00
Recv: OT prewarn	false	false	false	false
Recv: OT prewarn has
Recv: been triggered	false	false	false	false
Recv: off time		0	5	5	5
Recv: blank time	16	24	24	24
Recv: hysteresis
Recv: -end		-3	2	2	2
Recv: -start		1	3	3	3
Recv: Stallguard thrs				0
Recv: DRVSTATUS	X	Y	Z	E0
Recv: stallguard					X
Recv: sg_result					0
Recv: fsactive
Recv: stst			X	X	X
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:	X = 0x00:00:00:00
Recv: 	Y = 0xC0:07:00:00
Recv: 	Z = 0xC0:09:00:00
Recv: 	E0 = 0x81:08:00:00
```

**Comment by kakou-fr** [comment]
New test, i swap X with E0, now i can see X Y Z, but sometime, when i enter M122, one of the driver registers is bad (0x00 ...). Example with Y

```
Send: M122
Recv: 		X	Y	Z	E0
Recv: Enabled		true	true	true	false
Recv: Set current	500	500	600	550
Recv: RMS current	489	489	581	520
Recv: MAX current	689	689	819	733
Recv: Run current	15/31	15/31	18/31	16/31
Recv: Hold current	7/31	7/31	9/31	8/31
Recv: CS actual		7/31	7/31	9/31	8/31
Recv: PWM scale		50	50	104	1
Recv: vsense		1=.18	1=.18	1=.18	1=.18
Recv: stealthChop	true	true	true	true
Recv: msteps		16	16	16	16
Recv: tstep		1048575	1048575	1048575	1048575
Recv: pwm
Recv: threshold		79	79	329	1821
Recv: [mm/s]		100.08	100.08	6.01	1.00
Recv: OT prewarn	false	false	false	false
Recv: OT prewarn has
Recv: been triggered	false	false	false	false
Recv: off time		5	5	5	5
Recv: blank time	24	24	24	24
Recv: hysteresis
Recv: -end		2	2	2	2
Recv: -start		3	3	3	3
Recv: Stallguard thrs				0
Recv: DRVSTATUS	X	Y	Z	E0
Recv: stallguard
Recv: sg_result					0
Recv: fsactive
Recv: stst		X	X	X	X
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:	X = 0xC0:07:00:00
Recv: 	Y = 0xC0:07:00:00
Recv: 	Z = 0xC0:09:00:00
Recv: 	E0 = 0x80:08:00:00
Recv: 
Recv: 
Recv: ok
[...]
Send: M122
Recv: 		X	Y	Z	E0
Recv: Enabled		true	true	true	false
Recv: Set current	500	500	600	550
Recv: RMS current	489	489	581	520
Recv: MAX current	689	689	819	733
Recv: Run current	15/31	15/31	18/31	16/31
Recv: Hold current	7/31	7/31	9/31	8/31
Recv: CS actual		7/31	7/31	9/31	8/31
Recv: PWM scale		50	50	104	1
Recv: vsense		1=.18	1=.18	1=.18	1=.18
Recv: stealthChop	true	true	true	true
Recv: msteps		16	16	16	16
Recv: tstep		1048575	1048575	1048575	1048575
Recv: pwm
Recv: threshold		79	79	329	1821
Recv: [mm/s]		100.08	100.08	6.01	1.00
Recv: OT prewarn	false	false	false	false
Recv: OT prewarn has
Recv: been triggered	false	false	false	false
Recv: off time		5	5	5	5
Recv: blank time	24	24	24	24
Recv: hysteresis
Recv: -end		2	2	2	2
Recv: -start		3	3	3	3
Recv: Stallguard thrs				0
Recv: DRVSTATUS	X	Y	Z	E0
Recv: stallguard
Recv: sg_result					0
Recv: fsactive
Recv: stst		X	X	X	X
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:	X = 0xC0:07:00:00
Recv: 	Y = 0x00:00:00:00
Recv: 	Z = 0xC0:09:00:00
Recv: 	E0 = 0x80:08:00:00
```

**Comment by adriaanb** [comment]
Possibly a somewhat related issue for me - I get 3 out of 5 motors to work on my RUMBA board.
I chose the following pins:

Axis | TX | RX 
------------ | ------------- | -------------
X | D4 | D63 
Y | D5 | D64 
Z | D30| D14
E0 | D31 | D15 
E1 | D50 | D52 

I realize that D14 and D15 are the pins of an actual hardware UART on the Atmega2560, but according to the [documentation](https://www.arduino.cc/en/Reference/SoftwareSerial) they should work with the SoftwareSerial library.

The M122 comman nevertheless yields this:

```
SENDING:M122
		X	Y	Z	E0	E1
Enabled		true	true	false	false	true
Set current	900	900	800	700	700
RMS current	887	887	1436	1215	673
MAX current	1251	1251	2025	1713	949
Run current	28/31	28/31	25/31	21/31	21/31
Hold current	19/31	19/31	17/31	14/31	14/31
CS actual		19/31	19/31	0/31	0/31	14/31
PWM scale		22	22	0	0	16
vsense		1=.18	1=.18	0=.325	0=.325	1=.18
stealthChop	false	true	false	false	true
msteps		16	16	256	256	16
tstep		1048575	1048575	0	0	0
pwm
threshold		0	0	0	0	0
[mm/s]		-	-	-	-	-
OT prewarn	false	false	false	false	false
OT prewarn has
been triggered	false	false	false	false	false
off time		5	5	0	0	5
blank time	24	24	16	16	24
hysteresis
-end		2	2	-3	-3	2
-start		3	3	1	1	3
Stallguard thrs
DRVSTATUS	X	Y	Z	E0	E1
stst		X	X			X
olb
ola
s2gb
s2ga
otpw
ot
157C
150C
143C
120C
s2vsa
s2vsb
Driver registers:	X = 0xC0:13:00:00
	Y = 0xC0:13:00:00
	Z = 0x00:00:00:00
	E0 = 0x00:00:00:00
	E1 = 0xC0:0E:00:00
```

Just to see what would happen I disconnected the SD card reader of my RRD Full Graphics Smart Controller, disabled it in Configuration.h and connected Z to pins 30(TX) & 52(RX) and E0 to 15(TX) & 51(RX) (and moved E1 to pins 31 and 50) at which point I got data for all drivers (but cannot seem to get E0 to work/extrude reliably - which is probably a different issue on my end)

**Pull request merged**